0100-private-projects.md

```diff
@@ -0,0 +1,36 @@
+- Feature Name: Private Projects
+- Created: 2020-02-10
+- Associated Bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1602490
+
+# Summary
+
+Add an option to make a project "private", restricting who can see and interact with it.
+
+# Motivation
+
+It can be useful in some cases to be able to hide a project from users. For example, when a superuser creates a new project, they might want to make sure that everything is correctly set up before exposing it — running sync, verifying that it is enabled for the right locales, etc. In such a case, the project being not visible from non-superusers is enough. Once the superuser is done with configuration, they simply make the project public.
+
+For a different use-case, let's consider an agency working with several clients. They have a number of projects created on Pontoon, and they want to have specific localizers working on specific projects. It would be beneficial for them to able to restrict seeing a project to only a defined list of users. This requires being able to mark a project as private, and also having a configuration option to select a list of users who will be able to see, and interact, with that project. That would allow a clear separation of projects, reducing the risk of localizers translating the wrong set of projects.
+
+# Feature explanation
+
+When creating or editing a project, a superadmin has two new options:
+
+- a simple checkbox to make the project private or public;
+- a formset controlling which users can see the project when it is private.
+
+The "approved users" formset is visible only when the project is marked as private. It has two elements: a list of users, showing who currently has access (or will upon saving), and a search input, used to look for a specific user in the database and adding that user to the list. Clicking a user in the list removes that user from the list.
+
+Administrators and managers of a project always have access to it, regarless of whether they are in the approved users list or not.
+
+On each dashboard listing projects (`/projects/`, `/{locale}/`), private projects will not appear unless the user is authenticated and is in the list of approved users.
+
+Pages related to a project (`/{project}/`, `\{locale}/{project}/`, `/{locale}/{project}/{resource}/`) will return a 404 Not Found error unless the user is authenticated and is in the list of approved users. That should also be the case for back-end endpoints used to modify data, as a user should not be able to interact at all with a private project if they are not approved for that project.
```

Comment:
`/{project}/` -> `/projects/{project}/`

Reply:
We should also exclude private projects from permissions pages: `\{locale}/permissions/`.

Reply:
Private projects shouldn't be visible (for the non-authorized users) in the sync log too?

Reply:
I don't think it matters too much for sync log. I don't expect regular users to look at them (it's not even linked from anywhere in the site as far as I know), and it's not giving away too much information.

But as a side note, why are sync logs public? I'd expect them to be accessible only to superusers.

Reply:
We can limit access to sync log, but there's no real need for that as they aren't "giving away too much information". :)

Sometimes it's helpful it's public so that we can share links to particular sync logs in bugzilla.